src/libmongoc/src/mongoc/mongoc-stream-tls-openssl.c

```diff
@@ -631,14 +634,32 @@ _mongoc_stream_tls_openssl_handshake (mongoc_stream_t *stream,
 
    *events = 0;
 
-   if (!_mongoc_stream_tls_openssl_cert_verify_failed (ssl, error)) {
-      bson_set_error (error,
-                      MONGOC_ERROR_STREAM,
-                      MONGOC_ERROR_STREAM_SOCKET,
-                      "TLS handshake failed: %s",
-                      ERR_error_string (ERR_get_error (), NULL));
+   /* Try to relay certificate failure reason from OpenSSL library if any. */
+   if (_mongoc_stream_tls_openssl_cert_verify_failed (ssl, error)) {
+      RETURN (false);
    }
 
+   /* Otherwise, try to relay error info from OpenSSL. */
+   {
+      const unsigned long code = ERR_get_error ();
```

Comment:
Noted that [`ERR_get_error`](https://www.openssl.org/docs/man3.0/man3/ERR_get_error.html) pops an error code off of the queue, so you must capture its return value in lieu of calling it twice.